### ISSUE TITLE
Rule proposal: Disallow a class from implementing another class

### ISSUE THREAD

**Issue by cowwoc** [typescript-eslint/typescript-eslint#9370]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

If a class "implements" another class then `child instanceof parent` returns `false` which results in hard-to-catch bugs. I propose adding a rule that restricts the use of "implements" to interfaces and rejects it for abstract and concrete classes.

### Fail Cases

```TypeScript
interface ApplicationScope{};
class AbstractApplicationScope implements ApplicationScope{};
class MainApplicationScope implements AbstractApplicationScope{};

const value = new MainApplicationScope();
console.log(value instanceof AbstractApplicationScope); // returns false
```

### Pass Cases

```TypeScript
interface ApplicationScope{};
class AbstractApplicationScope implements ApplicationScope{};
class MainApplicationScope extends AbstractApplicationScope{};

const value = new MainApplicationScope();
console.log(value instanceof AbstractApplicationScope); // returns true
```

**Comment by kirkwaiblinger** [comment]
Thanks for the idea @cowwoc!

I'm tentatively +1 to this, though I sort of wonder whether it should just be an error in TS itself.

To add another example of where classes implementing classes is confusing, classes with non-public members _can't_ be implemented by other classes AFAICT. So, I could be wrong, but it seems almost _coincidentally permitted_ by TS that public-only classes can be implemented.

Have fun [playing around with this code](https://typescript-eslint.io/play/#ts=5.4.5&showAST=es&fileType=.tsx&code=FAYwNghgzlAEBiB7RsDexawA4CcCWAbhAC4CmsAZsgFywB2ArgLYBGpOA3MAL7CiQxYAIQg5YeJljCkmpOsThIU6TLkIlyVRLUat2XbkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false): 
```ts
class Foo {
  private foo: number;
}

class Bar implements Foo {
  private foo: number;
}
```

**Comment by bradzacher** [comment]
TS doesn't really model prototypal inheritance - it's all structural typing, baby.
Classes are sadly a glaring hole in TS and `instanceof` is quite unsound because of it. Ultimately TS just models `X instanceof Y` as "do all the properties of `Y` exist on `X`?"

For example:
```ts
class Foo { foo: number }
declare const x: { foo: number };
if (x instanceof Foo) {
    x;
//  ^? Foo 😢
}
```

Private members do change the equation slightly:
```ts
class Foo { private foo: number }
declare const x: { foo: number };
if (x instanceof Foo) {
    x;
//  ^? never ✅
}
```

However you can still do some fubar stuff eg:
```ts
class Foo { private foo: number }
declare const x: { };
if (x instanceof Foo) {
    x;
//  ^? Foo 😢
}

declare const y: { bar: number };
if (y instanceof Foo) {
    y;
//  ^? { bar: number } & Foo 😢
}

class Bar { private foo: number; bar: number }
if (y instanceof Bar) {
    y;
//  ^? Bar 😢
}
```

Classes are just sadly pretty unsound in TS because TS doesn't see a difference between an object and a class instance.

**Comment by cowwoc** [comment]
Doesn't this differ from the behavior of JS code? Isn't TS supposed to be a strict superset of JS?

**Comment by kirkwaiblinger** [comment]
Agreed that `instanceof` is not really modelled by the type system, so `instanceof`-correctness isn't really a motivation for me for having this rule. 

The motivation I see is that there's still no good reason I know of to ever `implements` a class. The implemented class (which may not even be in the same repo) may change at any time to have a private member, and then all code depending on implementing it will start to error. It seems like you'd only ever want to `extends` the class or `implements` an interface.

**Comment by bradzacher** [comment]
> The motivation I see is that there's still no good reason I know of to ever implements a class.

Dependency injection is one reason. "I want to provide an alternate implementation that matches the public API of the other class".

Using the class in the `implememts` clause means you can match the public API with enforcement in the class decl, as opposed to just at the usage.

But other than that yeah the cases are slim or you'd generally just swallow the duplication and extract the public signatures to an interface.

**Comment by kirkwaiblinger** [comment]
> Dependency injection is one reason. "I want to provide an alternate implementation that matches the public API of the other class".

I see. I'm not familiar with this pattern, but, even then, I'd think that this _only_ works if that other class has no protected/private members, now and forever, right....? 🤔 When would you actually have that guarantee? Unless the class is "pure abstract" in which case it's better expressed as an interface, anyway, right? In any case, maybe that's more of a "When not to use this rule" than a reason to not prohibit `implements`ing classes in the first place? Curious for your thoughts

**Comment by bradzacher** [comment]
It's a rarer usecase - more of a strawman argument really. In general you'd do it properly with a separate interface to define the expected API surface. Or in test usecases you'd use a mocking/spying library to do things.

I'd agree that this is something that should be filed upstream with TS. It does seem inconsistent that TS would disallow just the private/protected case.

**Comment by bradzacher** [comment]
> Doesn't this differ from the behavior of JS code? Isn't TS supposed to be a strict superset of JS?

TS **_syntax_** is a _rough_ superset of JS.
TS's type system is intended to roughly model how JS works - but it's far from perfect and makes a great deal more mistakes than people realise.

**Comment by kirkwaiblinger** [comment]
WAI on TS's end 
https://github.com/microsoft/TypeScript/issues/36449

**Comment by kirkwaiblinger** [comment]
Following https://stackoverflow.com/questions/38834625/extends-and-implements-difference-typescript, I guess you can write code like this....
```ts
class Parent {
  private parentPrivateMember: number;
}

class Child extends Parent {
  private childPrivateMember: number;
}

class GrandChild extends Child implements Parent {
  private grandChildPrivateMember: number;
}
```
In short, descendants of a class with private members are capable of `implements`ing it. As to _why_ you would want to do this... 🤷‍♂️

**Comment by bradzacher** [comment]
So if it's not gonna be fixed in TS then yeah I guess it's undesirable to an extent in a lot of cases.

I'm personally at a -1 in this living in our project:
- the incidences of this are rare and low-impact:
    - it's a pretty big typo to type `implements` instead of `extends`.
    - it should really only impact the `instanceof` operator as all type-level checks are done structurally so there's no type-level difference between the two (as shown above).
    - it only impacts classes with only public members.
- `implements` requires you to implement all members of the parent - which should be a red flag that you've done something wrong in the vast majority of usecases.
- you can't `override` the members of an implemented type (another red flag - esp if you have `noImplicitOverride` turned on).
- it will require type information so is limited to a subset of users.

For all of these reasons IMO it fails the "I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal)." requirement for new rules.

So yeah strong -1 from me.

**Comment by kirkwaiblinger** [comment]
Yeah, fair enough. I have encountered this in the wild before once, and it did cause a bit of a headache, but, yeah I can agree that it probably is not a common mistake, and therefore I can agree (reluctantly) that it's likely not worth the maintenance cost.

**Comment by cowwoc** [comment]
I personally ran into this use-case because I originally had interfaces as ancestors. Then I changed some of them to abstract classes, trying to make `instanceof` work but then I forgot to update the children classes, so they kept `implements`. I then wasted a couple of hours trying to figure out why `instanceof` was still not working.

In that respect (given that you never (?) want to `implements` a class) I still recommend adding such a rule and enabling it by default.

**Comment by bradzacher** [comment]
To be clear, @cowwoc, I'm not saying the rule wouldn't have value! We've identified that there is bad code that it could catch - it has value!

All I'm saying is that the occurence rate of the error (very low) combined with it requiring type info to catch (limiting the set of users it applies to) results in us estimating a low impact on our community, broadly speaking.
Put another way - if we add it to our recommended set tomorrow then we guess that there would be a very small number of reports across our entire userbase.

As volunteer maintainers with very limited bandwidth we need to weigh up the cost vs benefit of proposals. In this instance we don't believe that the value of the rule is high enough to justify the cost to our project.

You're welcome to build it yourself locally your your project and publish it as a 3rd party plugin! We encourage it even!
For now we shall be rejecting the proposal for our project. Thanks for filing!

**Comment by cowwoc** [comment]
No worries. I agree with your assessment. So long as the incidence rate remains low we might as well pass on this. Thanks.
